src/main/java/com/capitalone/dashboard/controller/SonarController.java

```diff
@@ -0,0 +1,145 @@
+package com.capitalone.dashboard.controller;
+
+import com.capitalone.dashboard.collector.SonarClient;
+import com.capitalone.dashboard.collector.SonarClientSelector;
+import com.capitalone.dashboard.model.CodeQuality;
+import com.capitalone.dashboard.model.CollectionError;
+import com.capitalone.dashboard.model.Collector;
+import com.capitalone.dashboard.model.SonarProject;
+import com.capitalone.dashboard.repository.CodeQualityRepository;
+import com.capitalone.dashboard.repository.CollectorRepository;
+import com.capitalone.dashboard.repository.SonarProjectRepository;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.simple.parser.ParseException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.client.HttpClientErrorException;
+
+import javax.validation.Valid;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+import static org.springframework.web.bind.annotation.RequestMethod.GET;
+
+@RestController
+@Validated
+public class SonarController {
+
+    private static final Log LOG = LogFactory.getLog(SonarController.class);
+
+    private final SonarProjectRepository sonarProjectRepository;
+    private final CodeQualityRepository codeQualityRepository;
+    private final CollectorRepository collectorRepository;
+    private SonarClientSelector sonarClientSelector;
+
+
+    private String instanceUrl;
+    private String projectName;
+    private String projectKey;
+    private Collector collector;
+    private SonarClient sonarClient;
+
+    @Autowired
+    public SonarController(SonarProjectRepository sonarProjectRepository,
+                           CodeQualityRepository codeQualityRepository,
+                           CollectorRepository collectorRepository,
+                           SonarClientSelector sonarClientSelector) {
+        this.sonarProjectRepository = sonarProjectRepository;
+        this.codeQualityRepository = codeQualityRepository;
+        this.collectorRepository = collectorRepository;
+        this.sonarClientSelector = sonarClientSelector;
+    }
+
+    @RequestMapping(value = "/refresh", method = GET, produces = APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> refresh(@Valid String projectName, @Valid String projectKey, @Valid String instanceUrl) {
```

Comment:
@nameisaravind  required params for sonar collector are projectName and instanceUrl, project key is not necessary here

Reply:
@nireeshT projectKey is optional value to support new sonar projects which are not exists in hygieia. Its one of the use case for pipeline testing tdm feature.